src/webhook/mutation/pod_mutator/oneagent_mutation/containers_test.go

```diff
@@ -150,10 +153,134 @@ func TestReinvokeUserContainers(t *testing.T) {
+			customBuildVersionAnnotationName: customBuildVersionValue,
+		}
+		namespaceAnnotations := map[string]string{
+			stageMappingAnnotationName: customStageFieldPath,
+			buildVersionAnnotationName: customBuildVersionFieldPath,
+		}
+		expectedMappings := map[string]string{
+			releaseVersionEnv:      defaultVersionLabelMapping[releaseVersionEnv],
+			releaseProductEnv:      defaultVersionLabelMapping[releaseProductEnv],
+			releaseStageEnv:        customStageFieldPath,
+			releaseBuildVersionEnv: customBuildVersionFieldPath,
+		}
+
+		doTestMappings(t, podAnnotations, namespaceAnnotations, expectedMappings, nil)
+	})
+	t.Run("all env vars are namespace-annotations driven", func(t *testing.T) {
+		podAnnotations := map[string]string{
+			customVersionAnnotationName:      customVersionValue,
+			customProductAnnotationName:      customProductValue,
+			customStageAnnotationName:        customReleaseStageValue,
+			customBuildVersionAnnotationName: customBuildVersionValue,
+		}
+		namespaceAnnotations := map[string]string{
+			versionMappingAnnotationName: customVersionFieldPath,
+			productMappingAnnotationName: customProductFieldPath,
+			stageMappingAnnotationName:   customStageFieldPath,
+			buildVersionAnnotationName:   customBuildVersionFieldPath,
+		}
+		expectedMappings := map[string]string{
+			releaseVersionEnv:      customVersionFieldPath,
+			releaseProductEnv:      customProductFieldPath,
+			releaseStageEnv:        customStageFieldPath,
+			releaseBuildVersionEnv: customBuildVersionFieldPath,
+		}
+
+		doTestMappings(t, podAnnotations, namespaceAnnotations, expectedMappings, nil)
+	})
+}
+
+func doTestMappings(t *testing.T, podAnnotations map[string]string, namespaceAnnotations map[string]string, expectedMappings map[string]string, unexpectedMappingsKeys []string) {
+	mutator := createTestPodMutator([]client.Object{getTestInitSecret()})
+	request := createTestMutationRequest(getTestComplexDynakube(), podAnnotations, getTestNamespace(namespaceAnnotations))
+	mutator.mutateUserContainers(request)
+
+	assertContainsMappings(t, expectedMappings, request)
+	assertDoesntContainMappings(t, unexpectedMappingsKeys, request)
+}
+
+func assertContainsMappings(t *testing.T, expectedMappings map[string]string, request *dtwebhook.MutationRequest) {
+	for envName, fieldPath := range expectedMappings {
+		assert.Contains(t, request.Pod.Spec.Containers[0].Env, corev1.EnvVar{
+			Name: envName,
+			ValueFrom: &corev1.EnvVarSource{
+				FieldRef: &corev1.ObjectFieldSelector{
+					APIVersion: "",
+					FieldPath:  fieldPath,
+				},
+			},
+		})
+	}
+}
+
+func assertDoesntContainMappings(t *testing.T, unexpectedMappingKeys []string, request *dtwebhook.MutationRequest) {
+	for _, key := range unexpectedMappingKeys {
+		idx := slices.IndexFunc(request.Pod.Spec.Containers[0].Env, func(envvar corev1.EnvVar) bool {
+			return envvar.Name == key
+		})
+		assert.Negative(t, idx, key+" found in container's envvars")
+	}
+}
```

Comment:
```suggestion
func assertNotContainsMappings(t *testing.T, unexpectedMappingKeys []string, request *dtwebhook.MutationRequest) {
	for _, env := range request.Pod.Spec.Containers[0].Env {
		assert.NotContains(t, unexpectedMappingKeys, env.Name)
	}
}
```

lets try not to use fancy experimental stuff, especially when there is a simpler solution

Reply:
Awesome, simpler == better!